---
title: "sgRNA target scoring"
author: "David Wu"
output: html_notebook
---
## Purpose
Assess off-targets of sgRNAs

## Setup 
Set working directory to project directory
```{r setup}
require(knitr)
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

### Load custom themes and functions
```{r, message=FALSE}
source('analysis/universal/themes.R') # themes for all project components

source('analysis/universal/functions.R') # themes for all project components
theme_set(theme_publication())

source('analysis/screen/scripts/functions.R')  # functions for this component; may be empty
```

### Directories
```{r, message=FALSE}
analysis_dir <- 'analysis/screen/output/sgrna_targets' # analysis file output directory
data_dir <- '' # data file output directory
figures_dir <- 'figures/screen'

dir.create(analysis_dir, showWarnings = FALSE)
dir.create(figures_dir, showWarnings = FALSE)
#dir.create(data_dir, showWarnings = FALSE)
```

### Load libraries
```{r, message=FALSE}
library(tidyverse)
library(rtracklayer)
library(plyranges)
library(patchwork)
library(extrafont)
library(Rsamtools)
```


### Annotation
```{r}
unified_reference <- import('analysis/reference/output/02_unified_reference/unified.gtf.gz')
tss_gtf <- import('analysis/screen/output/05_neighbors/screen_tss.gtf')
coding_sgrna <- read_tsv('analysis/screen/input/ScreenProcessing/library_tables/CRISPRi_v2_human_librarytable.txt')
lncrna_sgrna <- read_tsv('analysis/screen/input/ScreenProcessing/library_tables/CRISPRi_v2_human_lincRNA_unique_merged_librarytable.txt')
```
```{r}
chromosomes <- unified_reference %>% 
  as_tibble() %>% 
  select(feature_id, chromosome = seqnames) %>% 
  unique()
```




### sgRNA sequences
```{r}
sgrna_table <- coding_sgrna %>%
  filter(str_detect(sublibrary, 'top5')) %>% 
  separate(sgId, into = c('feature_id', 'sense', 'position'), sep = '_', remove = F) %>% 
  separate(position, into = c('position', NA, NA)) %>% 
  bind_rows(
    lncrna_sgrna %>% 
  separate(`...1`, into = c(NA, 'sgId'), sep = ':') %>% 
  separate(sgId, into = c(NA, 'sense', 'position', NA), sep = '_|\\.', remove = F) %>% 
  select(feature_id = gene, everything())
  ) %>% 
  mutate(pam = 'NGG') %>% 
  select(feature_id, sgId, sequence, pam, position, sense) %>% 
  filter(!str_detect(sgId, 'non-targeting'))

sgrna_table %>% tail()
```

### Load screen data
```{r}
neighbor_wide <- read_tsv('analysis/screen/output/05_neighbors/neighbor_wide.tsv.gz') %>% 
  mutate(primary_pheno = factor(primary_pheno, levels = c('Differentiation', 'Dual', 'Proliferation', 'None')),
         status = ifelse(primary_pheno == 'None', 'Non-hit', 'Hit')) 

neighbor_wide
```

### Extract hit sgRNAs
```{r}
hits <- neighbor_wide %>% 
  filter(diff_full_status == 'Hit' | prol_full_status == 'Hit',
         locus_ambiguity == 'Non-ambiguous' | library == 'Coding')

hits
```
```{r}
sgrnas <- sgrna_table %>% 
  filter(feature_id %in% hits$feature_id) %>% 
  left_join(chromosomes) %>% 
  mutate(sgId = str_replace_all(sgId, ',', ';')) %>% 
  mutate(sequence = str_replace(sequence, 'G', 'N')) %>% 
  unique()

sgrnas
```
```{r}
sgrnas %>% 
  select(id = sgId, sequence, pam, chromosome, position, sense) %>% 
  write_csv(file.path(analysis_dir, 'kmers.csv'))
```

```{bash eval=FALSE}
INDEX=/home/genomes/Homo_sapiens/Ensembl/GRCh37/release74/guidescan2/grch37
FASTA=/home/genomes/Homo_sapiens/Ensembl/GRCh37/release74/genome/GRCh37.p13.genome.fa
KMERS=/media/data4/shared/ultrawide/analysis/screen/output/sgrna_targets/kmers.csv
guidescan enumerate $INDEX -f $KMERS -n 1 -o db.sam -a NGG

python /home/software/guidescan-cli/scripts/decode_database.py db.sam $FASTA > processed.csv

```

```{r}
sam <- read_tsv('data/guidescan/db.sam', comment = '@', col_names = c('id', 'flag', 'chromosome', 'position', 'mapq', 'cigar', 'rnext', 'pnext', 'tlen', 'sequence', 'quality')) 
sam %>% 
  arrange(chromosome, position) %>% 
  add_count(id) %>% 
  arrange(-n, id)

```
```{r}
sam %>% 
  filter(id == 'AMER1_+_63425562.23-P1P2')
```
## Export
```{r}

```
